### PR TITLE
Adding ReasonForError usage:

### PR DESCRIPTION
**Pull request by acornies**
To address #362 by mapping more appropriate http error responses for secrets API.

## Description
- Required update k8s.io/apimachinery to 1.9
- Added processErrorReasons mapping private function

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

## How Has This Been Tested?
TODO

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: handlers/secrets.go [hunk 1]
```diff
@@ -54,8 +54,7 @@ func getSecretsHandler(kube typedV1.SecretInterface, w http.ResponseWriter, r *h
 	selector := fmt.Sprintf("%s=%s", secretLabel, secretLabelValue)
 	res, err := kube.List(metav1.ListOptions{LabelSelector: selector})
 	if err != nil {
-		w.WriteHeader(http.StatusInternalServerError)
-		log.Printf("Secrets query error: %v\n", err)
```
Comment (alexellis) [inline handlers/secrets.go]: Why did we drop the logging?

Comment (acornies) [reply handlers/secrets.go]: it's only been moved.

Comment (acornies) [reply handlers/secrets.go]: actually, going to update it

**Comment by alexellis** [comment]
Is CI blocked? Do we know why?

**Comment by LucasRoesler** [comment]
I don't see any link for CI details, which is really weird

**Comment by alexellis** [comment]
Why is this PR draft? Is that why? Is this WIP or ready for review @acornies ?

File: handlers/secrets.go [hunk 2]
```diff
@@ -311,3 +307,27 @@ func removeVolumeMount(volumeName string, mounts []apiv1.VolumeMount) []apiv1.Vo
...
+	case k8serrors.ReasonForError(err) == metav1.StatusReasonConflict:
+		log.Printf("Secret %s error reason: %s, %v\n", context, metav1.StatusReasonConflict, err)
+		return http.StatusConflict
+	case k8serrors.ReasonForError(err) == metav1.StatusReasonInvalid:
+		log.Printf("Secret %s error reason: %s, %v\n", context, metav1.StatusReasonInvalid, err)
+		return http.StatusUnprocessableEntity
+	case k8serrors.ReasonForError(err) == metav1.StatusReasonBadRequest:
+		log.Printf("Secret %s error reason: %s, %v\n", context, metav1.StatusReasonBadRequest, err)
+		return http.StatusBadRequest
+	case k8serrors.ReasonForError(err) == metav1.StatusReasonForbidden:
```
Comment (alexellis) [inline handlers/secrets.go]: I guess my concern with this approach is that we don't give the reason for the error back to the caller. 

We should also be able to test this new method that has been broken out with unit tests when we have a final design on how it is going to work. 

I am not super keen on logging in the function. I'd rather see you return the Status Code plus a message that we then chose to log in the calling function - thereby applying Single Responsibility Principle (SRP)

**Comment by LucasRoesler** [comment]
The last build on travis looks like this https://travis-ci.org/openfaas/faas-cli/builds/504025019

for 54ee31dd7e3d538783ba9466f81dc5ab98a793bd

**Comment by acornies** [comment]
Yeah, the draft PRs mess with CI usually. I chose it so that we could deliberate on the change first (pre-unit tests as well). I'll update with recent feedback.

**Comment by acornies** [comment]
I've updated the PR. Can someone flip this PR status to non-draft mode? I don't have permissions.

File: handlers/secrets.go (hunk 2 above)
Comment (acornies) [reply handlers/secrets.go]: Updated.

**Comment by alexellis** [comment]
I would have thought since it's your PR you can set it out of draft. What do the GH docs say? This is a new feature and I don't entirely see the point of it

**Comment by acornies** [comment]
Normally you can: https://help.github.com/en/articles/changing-the-stage-of-a-pull-request. Is it because I'm not a member of any teams?

**Pull request merged**